File: ui/src/stores/params.ts
import { defineStore } from 'pinia'
import { format } from 'date-fns'
import { mande } from 'mande'
import cloneDeep from 'lodash.clonedeep'


const params = mande('/')

export interface ParamState {
    Categories: Array<Category>;
    Statuses: Array<Status>;
    DeviceTypes: Array<DeviceType>;
    InterfaceTypes: Array<InterfaceType>;
    Locations: Array<Location>;
    Maintainers: Array<Maintainer>;
    OperatingSystems: Array<OperatingSystem>;
    Architectures: Array<Architecture>;
    VLANs: Array<VLAN>;
    Users: Array<User>;
}

export interface Category {
    ID: number | null;
    Label: string;
    Icon: string;
    Notes: string | null;
    IsLocked: boolean;

}
export interface Status {
    ID: number | null;
    Label: string;
    Icon: string;
    Notes: string | null;
    IsLocked: boolean;
}
export interface DeviceType {
    ID: number | null;
    Label: string;
    Icon: string;
    Notes: string | null;
    IsLocked: boolean;
}
export interface InterfaceType {
    ID: number | null;
    Label: string;
    Icon: string;
    Notes: string | null;
    IsLocked: boolean;
}
export interface Location {
    ID: number | null;
    Label: string;
    IsCloud: number;
    Notes: string | null;
    IsLocked: boolean;
}
export interface Maintainer {
    ID: number | null;
    Label: string;
    IsInternal: number;
    Notes: string | null;
    IsLocked: boolean;
}
export interface OperatingSystem {
    ID: number | null;
    Vendor: string;
    Family: string;
    Name: string;
    Version: string;
    IsOpenSource: boolean;
    IsServer: boolean;
    Notes: string | null;
    IsLocked: boolean;
}
export interface Architecture {
    ID: number | null;
    Label: string;
    BitSpace: number;
    Notes: string | null;
    IsLocked: boolean;
}
export interface VLAN {
    ID: number | null;
    Label: string;
    IPv4Mask: string;
    IPv6Mask: string;
    Notes: string | null;
    IsLocked: boolean;
}
export interface User {
    ID: number | null;
    Label: string;
    Password: string;
    NewPassword: string | null;
    Access: string;
    Notes: string | null;
    IsLocked: boolean;
}


export const useParams = defineStore('params', {
    state: () => {
        return ({
            Categories: [],
            Statuses: [],
            DeviceTypes: [],
            InterfaceTypes: [],
            Locations: [],
            Maintainers: [],
            OperatingSystems: [],
            Architectures: [],
            VLANs: [],
            Users: [],

        } as ParamState)
    },
    actions: {
        getParams() {
            params.get<ParamState>('/params').then((response) => {
                Object.keys(response).forEach((key) => {
                    switch (key) {
                        case 'Categories':
                            this.Categories = response.Categories
                            break
                        case 'Statuses':
                            this.Statuses = response.Statuses
                            break
                        case 'DeviceTypes':
                            this.DeviceTypes = response.DeviceTypes
                            break
                        case 'InterfaceTypes':
                            this.InterfaceTypes = response.InterfaceTypes
                            break
                        case 'Maintainers':
                            this.Maintainers = response.Maintainers
                            break
                        case 'Locations':
                            this.Locations = response.Locations
                            break
                        case 'OperatingSystems':
                            this.OperatingSystems = response.OperatingSystems
                            break
                        case 'Architectures':
                            this.Architectures = response.Architectures
                            break
                        case 'VLANs':
                            this.VLANs = response.VLANs
                            break
                        case 'Users':
                            this.Users = response.Users
                            break
                    }
                })
            })
        },

        NewCategory() {
            this.Categories.splice(0, 0, { ID: null, Label: "", Icon: "", Notes: null, IsLocked: false })
        },
        SaveCategory(index: number) {
            const isNew = ((this.Categories[index].ID === null) || (this.Categories[index].ID === 0))
            params.post('/categories', this.Categories[index]).then((response: any) => {
                this.Categories[index].ID = response.ID
                if (isNew) {
                    const move = this.Categories.splice(index, 1)
                    move.forEach((item => {
                        this.Categories.push(item)
                    }))
                }
            })
        },
        DropCategory(index: number) {
            this.Categories.splice(index, 1)
        },
        DeleteCategory(index: number) {
            params.delete('/categories/' + this.Categories[index].ID).then((response: any) => {
                this.Categories.splice(index, 1)
            })
        },

        NewStatus() {
            this.Statuses.splice(0, 0, { ID: null, Label: "", Icon: "", Notes: null, IsLocked: false })
        },
        SaveStatus(index: number) {
            const isNew = ((this.Statuses[index].ID === null) || (this.Statuses[index].ID === 0))
            params.post('/statuses', this.Statuses[index]).then((response: any) => {
                this.Statuses[index].ID = response.ID
                if (isNew) {
                    const move = this.Statuses.splice(index, 1)
                    move.forEach((item => {
                        this.Statuses.push(item)
                    }))
                }
            })
        },
        DropStatus(index: number) {
            this.Statuses.splice(index, 1)
        },
        DeleteStatus(index: number) {
            params.delete('/statuses/' + this.Statuses[index].ID).then((response: any) => {
                this.Statuses.splice(index, 1)
            })
        },

        NewMaintainer() {
            this.Maintainers.splice(0, 0, { ID: null, Label: "", Notes: null, IsInternal: 0, IsLocked: false })
        },
        SaveMaintainer(index: number) {
            const isNew = ((this.Maintainers[index].ID === null) || (this.Maintainers[index].ID === 0))

            params.post('/maintainers', this.Maintainers[index]).then((response: any) => {
                this.Maintainers[index].ID = response.ID
                if (isNew) {
                    const move = this.Maintainers.splice(index, 1)
                    move.forEach((item => {
                        this.Maintainers.push(item)
                    }))
                }
            })
        },
        DropMaintainer(index: number) {
            this.Maintainers.splice(index, 1)
        },
        DeleteMaintainer(index: number) {
            params.delete('/maintainers/' + this.Maintainers[index].ID).then((response: any) => {
                this.Maintainers.splice(index, 1)
            })
        },

        NewLocation() {
            this.Locations.splice(0, 0, { ID: null, Label: "", IsCloud: 0, Notes: null, IsLocked: false })
        },
        SaveLocation(index: number) {
            const isNew = ((this.Locations[index].ID === null) || (this.Locations[index].ID === 0))

            params.post('/locations', this.Locations[index]).then((response: any) => {
                this.Locations[index].ID = response.ID
                if (isNew) {
                    const move = this.Locations.splice(index, 1)
                    move.forEach((item => {
                        this.Locations.push(item)
                    }))
                }
            })
        },
        DropLocation(index: number) {
            this.Locations.splice(index, 1)
        },
        DeleteLocation(index: number) {
            params.delete('/locations/' + this.Locations[index].ID).then((response: any) => {
                this.Locations.splice(index, 1)
            })
        },

        NewDeviceType() {
            this.DeviceTypes.splice(0, 0, { ID: null, Label: "", Icon: "", Notes: null, IsLocked: false })
        },
        SaveDeviceType(index: number) {
            const isNew = ((this.DeviceTypes[index].ID === null) || (this.DeviceTypes[index].ID === 0))


            params.post('/deviceTypes', this.DeviceTypes[index]).then((response: any) => {
                this.DeviceTypes[index].ID = response.ID
                if (isNew) {
                    const move = this.DeviceTypes.splice(index, 1)
                    move.forEach((item => {
                        this.DeviceTypes.push(item)
                    }))
                }
            })
        },
        DropDeviceType(index: number) {
            this.DeviceTypes.splice(index, 1)
        },
        DeleteDeviceType(index: number) {
            params.delete('/deviceTypes/' + this.DeviceTypes[index].ID).then((response: any) => {
                this.DeviceTypes.splice(index, 1)
            })
        },

        NewOperatingSystem() {
            this.OperatingSystems.splice(0, 0, { ID: null, Vendor: "", Family: "", Name: "", Version: "", IsOpenSource: false, IsServer: false, Notes: null, IsLocked: false })
        },
        SaveOperatingSystem(index: number) {
            const isNew = ((this.OperatingSystems[index].ID === null) || (this.OperatingSystems[index].ID === 0))
            params.post('/operatingSystems', this.OperatingSystems[index]).then((response: any) => {
                this.OperatingSystems[index].ID = response.ID
                if (isNew) {
                    const move = this.OperatingSystems.splice(index, 1)
                    move.forEach((item => {
                        this.OperatingSystems.push(item)
                    }))
                }
            })
        },
        DropOperatingSystem(index: number) {
            this.OperatingSystems.splice(index, 1)
        },
        DeleteOperatingSystem(index: number) {
            params.delete('/operatingSystems/' + this.OperatingSystems[index].ID).then((response: any) => {
                this.OperatingSystems.splice(index, 1)
            })
        },

        NewArchitecture() {
            this.Architectures.splice(0, 0, { ID: null, Label: "", BitSpace: 0, Notes: null, IsLocked: false })
        },
        SaveArchitecture(index: number) {
            const isNew = ((this.Architectures[index].ID === null) || (this.Architectures[index].ID === 0))


            params.post('/architectures', this.Architectures[index]).then((response: any) => {
                this.Architectures[index].ID = response.ID
                if (isNew) {
                    const move = this.Architectures.splice(index, 1)
                    move.forEach((item => {
                        this.Architectures.push(item)
                    }))
                }
            })
        },
        DropArchitecture(index: number) {
            this.Architectures.splice(index, 1)
        },
        DeleteArchitecture(index: number) {
            params.delete('/architectures/' + this.Architectures[index].ID).then((response: any) => {
                this.Architectures.splice(index, 1)
            })
        },

        NewVLAN() {
            this.VLANs.splice(0, 0, { ID: null, Label: "", IPv4Mask: "", IPv6Mask: "", Notes: null, IsLocked: false })
        },
        SaveVLAN(index: number) {
            const isNew = ((this.VLANs[index].ID === null) || (this.VLANs[index].ID === 0))


            params.post('/vlans', this.VLANs[index]).then((response: any) => {
                this.VLANs[index].ID = response.ID
                if (isNew) {
                    const move = this.VLANs.splice(index, 1)
                    move.forEach((item => {
                        this.VLANs.push(item)
                    }))
                }
            })
        },
        DropVLAN(index: number) {
            this.VLANs.splice(index, 1)
        },
        DeleteVLAN(index: number) {
            params.delete('/vlans/' + this.VLANs[index].ID).then((response: any) => {
                this.VLANs.splice(index, 1)
            })
        },

        NewUser() {
            this.Users.splice(0, 0, { ID: null, Label: "", NewPassword: null, Password: "", Access: "", Notes: null, IsLocked: false })
        },
        SaveUser(index: number) {
            const isNew = ((this.Users[index].ID === null) || (this.Users[index].ID === 0))


            params.post('/users', this.Users[index]).then((response: any) => {
                this.Users[index].ID = response.ID
                if (isNew) {
                    const move = this.Users.splice(index, 1)
                    move.forEach((item => {
                        this.Users.push(item)
                    }))
                }
            })
        },
        DropUser(index: number) {
            this.Users.splice(index, 1)
        },
        DeleteUser(index: number) {
            params.delete('/users/' + this.Users[index].ID).then((response: any) => {
                this.Users.splice(index, 1)
            })
        },

    }
})

File: api/handlers/params/params.go
package params

import (
	"net/http"
	"strconv"

	"facemasq/lib/db"
	"facemasq/lib/formats"
	"facemasq/lib/logging"
	"facemasq/lib/translate"
	"facemasq/models"

	"github.com/gorilla/mux"
)

type Params struct {
	Categories       []models.Category
	Statuses         []models.Status
	Locations        []models.Location
	Maintainers      []models.Maintainer
	Architectures    []models.Architecture
	OperatingSystems []models.OperatingSystem
	InterfaceTypes   []models.InterfaceType
	DeviceTypes      []models.DeviceType
	VLANs            []models.VLAN
	Users            []models.User
}

func GetAll(out http.ResponseWriter, in *http.Request) {
	var params Params
	var err error
	err = db.Conn.NewSelect().Model(&params.Categories).Scan(db.Context)
	if err != nil {
		http.Error(out, translate.Message("RetrieveCategoryError", "Unable to retrieve Category data"), http.StatusInternalServerError)
	}

	err = db.Conn.NewSelect().Model(&params.Statuses).Scan(db.Context)
	if err != nil {
		http.Error(out, translate.Message("RetrieveStatusError", "Unable to retrieve Status data"), http.StatusInternalServerError)
	}

	err = db.Conn.NewSelect().Model(&params.Locations).Scan(db.Context)
	if err != nil {
		http.Error(out, translate.Message("RetrieveLocationError", "Unable to retrieve Location data"), http.StatusInternalServerError)
	}

	err = db.Conn.NewSelect().Model(&params.Maintainers).Scan(db.Context)
	if err != nil {
		http.Error(out, translate.Message("RetrieveMaintainerError", "Unable to retrieve Maintainer data"), http.StatusInternalServerError)
	}

	err = db.Conn.NewSelect().Model(&params.Architectures).Scan(db.Context)
	if err != nil {
		http.Error(out, translate.Message("RetrieveArchitectureError", "Unable to retrieve Architecture data"), http.StatusInternalServerError)
	}

	err = db.Conn.NewSelect().Model(&params.OperatingSystems).Scan(db.Context)
	if err != nil {
		http.Error(out, translate.Message("RetrieveOperatingSystemError", "Unable to retrieve OperatingSystem data"), http.StatusInternalServerError)
	}

	err = db.Conn.NewSelect().Model(&params.InterfaceTypes).Scan(db.Context)
	if err != nil {
		http.Error(out, translate.Message("RetrieveInterfaceTypeError", "Unable to retrieve InterfaceType data"), http.StatusInternalServerError)
	}

	err = db.Conn.NewSelect().Model(&params.DeviceTypes).Scan(db.Context)
	if err != nil {
		http.Error(out, translate.Message("RetrieveDeviceTypeError", "Unable to retrieve DeviceType data"), http.StatusInternalServerError)
	}

	err = db.Conn.NewSelect().Model(&params.VLANs).Scan(db.Context)
	if err != nil {
		http.Error(out, translate.Message("RetrieveVLANError", "Unable to retrieve VLAN data"), http.StatusInternalServerError)
	}

	err = db.Conn.NewSelect().Model(&params.Users).Scan(db.Context)
	if err != nil {
		http.Error(out, translate.Message("RetrieveUserError", "Unable to retrieve User data"), http.StatusInternalServerError)
	}

	formats.WriteJSONResponse(params, out, in)
}

func SaveCategory(out http.ResponseWriter, in *http.Request) {
	var input models.Category
	err := formats.ReadJSONBody(in, &input)
	if err != nil {
		translation := translate.Message("SaveCategoryError", "Unable to save Category")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return
	}

	if input.ID > 0 {
		_, err = db.Conn.NewUpdate().Model(&input).Where(`id = ?`, input.ID).Exec(db.Context)
	} else {
		_, err = db.Conn.NewInsert().Model(&input).Exec(db.Context)
	}
	if err != nil {
		translation := translate.Message("SaveCategoryError", "Unable to save Category")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return

	}
	formats.WriteJSONResponse(input, out, in)
}

func DeleteCategory(out http.ResponseWriter, in *http.Request) {
	var input models.Category
	id, err := strconv.ParseInt(mux.Vars(in)["ID"], 10, 64)
	if err != nil {
		translation := translate.Message("DeleteCategory", "Unable to delete Category")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return

	}
	input.ID = id

	_, err = db.Conn.NewDelete().Model(&input).WherePK().Exec(db.Context)
	if err != nil {
		translation := translate.Message("DeleteCategory", "Unable to delete Category")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return
	}
	formats.WriteJSONResponse(input, out, in)
}

func SaveStatus(out http.ResponseWriter, in *http.Request) {
	var input models.Status
	err := formats.ReadJSONBody(in, &input)
	if err != nil {
		translation := translate.Message("SaveStatusError", "Unable to save Status")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return
	}

	if input.ID > 0 {
		_, err = db.Conn.NewUpdate().Model(&input).Where(`id = ?`, input.ID).Exec(db.Context)
	} else {
		_, err = db.Conn.NewInsert().Model(&input).Exec(db.Context)
	}
	if err != nil {
		translation := translate.Message("SaveStatusError", "Unable to save Status")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return

	}
	formats.WriteJSONResponse(input, out, in)
}

func DeleteStatus(out http.ResponseWriter, in *http.Request) {
	var input models.Status
	id, err := strconv.ParseInt(mux.Vars(in)["ID"], 10, 64)
	if err != nil {
		translation := translate.Message("DeleteStatus", "Unable to delete Status")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return

	}
	input.ID = id

	_, err = db.Conn.NewDelete().Model(&input).WherePK().Exec(db.Context)
	if err != nil {
		translation := translate.Message("DeleteStatus", "Unable to delete Status")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return
	}
	formats.WriteJSONResponse(input, out, in)
}

func SaveLocation(out http.ResponseWriter, in *http.Request) {
	var input models.Location
	err := formats.ReadJSONBody(in, &input)
	if err != nil {
		translation := translate.Message("SaveLocationError", "Unable to save Location")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return
	}

	if input.ID > 0 {
		_, err = db.Conn.NewUpdate().Model(&input).Where(`id = ?`, input.ID).Exec(db.Context)
	} else {
		_, err = db.Conn.NewInsert().Model(&input).Exec(db.Context)
	}
	if err != nil {
		translation := translate.Message("SaveLocationError", "Unable to save Location")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return

	}
	formats.WriteJSONResponse(input, out, in)
}

func DeleteLocation(out http.ResponseWriter, in *http.Request) {
	var input models.Location
	id, err := strconv.ParseInt(mux.Vars(in)["ID"], 10, 64)
	if err != nil {
		translation := translate.Message("DeleteLocation", "Unable to delete Location")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return

	}
	input.ID = id

	_, err = db.Conn.NewDelete().Model(&input).WherePK().Exec(db.Context)
	if err != nil {
		translation := translate.Message("DeleteLocation", "Unable to delete Location")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return
	}
	formats.WriteJSONResponse(input, out, in)
}

func SaveMaintainer(out http.ResponseWriter, in *http.Request) {
	var input models.Maintainer
	err := formats.ReadJSONBody(in, &input)
	if err != nil {
		translation := translate.Message("SaveMaintainerError", "Unable to save Maintainer")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return
	}

	if input.ID > 0 {
		_, err = db.Conn.NewUpdate().Model(&input).Where(`id = ?`, input.ID).Exec(db.Context)
	} else {
		_, err = db.Conn.NewInsert().Model(&input).Exec(db.Context)
	}
	if err != nil {
		translation := translate.Message("SaveMaintainerError", "Unable to save Maintainer")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return

	}
	formats.WriteJSONResponse(input, out, in)
}

func DeleteMaintainer(out http.ResponseWriter, in *http.Request) {
	var input models.Maintainer
	id, err := strconv.ParseInt(mux.Vars(in)["ID"], 10, 64)
	if err != nil {
		translation := translate.Message("DeleteMaintainer", "Unable to delete Maintainer")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return

	}
	input.ID = id

	_, err = db.Conn.NewDelete().Model(&input).WherePK().Exec(db.Context)
	if err != nil {
		translation := translate.Message("DeleteMaintainer", "Unable to delete Maintainer")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return
	}
	formats.WriteJSONResponse(input, out, in)
}

func SaveArchitecture(out http.ResponseWriter, in *http.Request) {
	var input models.Architecture
	err := formats.ReadJSONBody(in, &input)
	if err != nil {
		translation := translate.Message("SaveArchitectureError", "Unable to save Architecture")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return
	}
	if input.ID > 0 {
		_, err = db.Conn.NewUpdate().Model(&input).Where(`id = ?`, input.ID).Exec(db.Context)
	} else {
		_, err = db.Conn.NewInsert().Model(&input).Exec(db.Context)
	}
	if err != nil {
		translation := translate.Message("SaveArchitectureError", "Unable to save Architecture")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return

	}
	formats.WriteJSONResponse(input, out, in)
}

func DeleteArchitecture(out http.ResponseWriter, in *http.Request) {
	var input models.Architecture
	id, err := strconv.ParseInt(mux.Vars(in)["ID"], 10, 64)
	if err != nil {
		translation := translate.Message("DeleteArchitecture", "Unable to delete Architecture")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return

	}
	input.ID = id

	_, err = db.Conn.NewDelete().Model(&input).WherePK().Exec(db.Context)
	if err != nil {
		translation := translate.Message("DeleteArchitecture", "Unable to delete Architecture")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return
	}
	formats.WriteJSONResponse(input, out, in)
}

func SaveOperatingSystem(out http.ResponseWriter, in *http.Request) {
	var input models.OperatingSystem
	err := formats.ReadJSONBody(in, &input)
	if err != nil {
		translation := translate.Message("SaveOperatingSystemError", "Unable to save OperatingSystem")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return
	}

	if input.ID > 0 {
		_, err = db.Conn.NewUpdate().Model(&input).Where(`id = ?`, input.ID).Exec(db.Context)
	} else {
		_, err = db.Conn.NewInsert().Model(&input).Exec(db.Context)
	}
	if err != nil {
		translation := translate.Message("SaveOperatingSystemError", "Unable to save OperatingSystem")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return

	}
	formats.WriteJSONResponse(input, out, in)
}

func DeleteOperatingSystem(out http.ResponseWriter, in *http.Request) {
	var input models.OperatingSystem
	id, err := strconv.ParseInt(mux.Vars(in)["ID"], 10, 64)
	if err != nil {
		translation := translate.Message("DeleteOperatingSystem", "Unable to delete OperatingSystem")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return

	}
	input.ID = id

	_, err = db.Conn.NewDelete().Model(&input).WherePK().Exec(db.Context)
	if err != nil {
		translation := translate.Message("DeleteOperatingSystem", "Unable to delete OperatingSystem")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return
	}
	formats.WriteJSONResponse(input, out, in)
}

func SaveInterfaceType(out http.ResponseWriter, in *http.Request) {
	var input models.InterfaceType
	err := formats.ReadJSONBody(in, &input)
	if err != nil {
		translation := translate.Message("SaveInterfaceTypeError", "Unable to save InterfaceType")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return
	}

	if input.ID > 0 {
		_, err = db.Conn.NewUpdate().Model(&input).Where(`id = ?`, input.ID).Exec(db.Context)
	} else {
		_, err = db.Conn.NewInsert().Model(&input).Exec(db.Context)
	}
	if err != nil {
		translation := translate.Message("SaveInterfaceTypeError", "Unable to save InterfaceType")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return

	}
	formats.WriteJSONResponse(input, out, in)
}

func DeleteInterfaceType(out http.ResponseWriter, in *http.Request) {
	var input models.InterfaceType
	id, err := strconv.ParseInt(mux.Vars(in)["ID"], 10, 64)
	if err != nil {
		translation := translate.Message("DeleteInterfaceType", "Unable to delete InterfaceType")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return

	}
	input.ID = id

	_, err = db.Conn.NewDelete().Model(&input).WherePK().Exec(db.Context)
	if err != nil {
		translation := translate.Message("DeleteInterfaceType", "Unable to delete InterfaceType")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return
	}
	formats.WriteJSONResponse(input, out, in)
}

func SaveDeviceType(out http.ResponseWriter, in *http.Request) {
	var input models.DeviceType
	err := formats.ReadJSONBody(in, &input)
	if err != nil {
		translation := translate.Message("SaveDeviceTypeError", "Unable to save DeviceType")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return
	}

	if input.ID > 0 {
		_, err = db.Conn.NewUpdate().Model(&input).Where(`id = ?`, input.ID).Exec(db.Context)
	} else {
		_, err = db.Conn.NewInsert().Model(&input).Exec(db.Context)
	}
	if err != nil {
		translation := translate.Message("SaveDeviceTypeError", "Unable to save DeviceType")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return

	}
	formats.WriteJSONResponse(input, out, in)
}

func DeleteDeviceType(out http.ResponseWriter, in *http.Request) {
	var input models.DeviceType
	id, err := strconv.ParseInt(mux.Vars(in)["ID"], 10, 64)
	if err != nil {
		translation := translate.Message("DeleteDeviceType", "Unable to delete DeviceType")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return

	}
	input.ID = id

	_, err = db.Conn.NewDelete().Model(&input).WherePK().Exec(db.Context)
	if err != nil {
		translation := translate.Message("DeleteDeviceType", "Unable to delete DeviceType")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return
	}
	formats.WriteJSONResponse(input, out, in)
}

func SaveVLAN(out http.ResponseWriter, in *http.Request) {
	var input models.VLAN
	err := formats.ReadJSONBody(in, &input)
	if err != nil {
		translation := translate.Message("SaveVLANError", "Unable to save VLAN")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return
	}

	if input.ID > 0 {
		_, err = db.Conn.NewUpdate().Model(&input).Where(`id = ?`, input.ID).Exec(db.Context)
	} else {
		_, err = db.Conn.NewInsert().Model(&input).Exec(db.Context)
	}
	if err != nil {
		translation := translate.Message("SaveVLANError", "Unable to save VLAN")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return

	}
	formats.WriteJSONResponse(input, out, in)
}

func DeleteVLAN(out http.ResponseWriter, in *http.Request) {
	var input models.VLAN
	id, err := strconv.ParseInt(mux.Vars(in)["ID"], 10, 64)
	if err != nil {
		translation := translate.Message("DeleteVLAN", "Unable to delete VLAN")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return

	}
	input.ID = id

	_, err = db.Conn.NewDelete().Model(&input).WherePK().Exec(db.Context)
	if err != nil {
		translation := translate.Message("DeleteVLAN", "Unable to delete VLAN")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return
	}
	formats.WriteJSONResponse(input, out, in)
}

func SaveUser(out http.ResponseWriter, in *http.Request) {
	var input models.User
	err := formats.ReadJSONBody(in, &input)
	if err != nil {
		translation := translate.Message("SaveUserError", "Unable to save User")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return
	}

	if input.ID > 0 {
		_, err = db.Conn.NewUpdate().Model(&input).Where(`id = ?`, input.ID).Exec(db.Context)
	} else {
		_, err = db.Conn.NewInsert().Model(&input).Exec(db.Context)
	}
	if err != nil {
		translation := translate.Message("SaveUserError", "Unable to save User")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return

	}
	formats.WriteJSONResponse(input, out, in)
}

func DeleteUser(out http.ResponseWriter, in *http.Request) {
	var input models.User
	id, err := strconv.ParseInt(mux.Vars(in)["ID"], 10, 64)
	if err != nil {
		translation := translate.Message("DeleteUser", "Unable to delete User")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return

	}
	input.ID = id

	_, err = db.Conn.NewDelete().Model(&input).WherePK().Exec(db.Context)
	if err != nil {
		translation := translate.Message("DeleteUser", "Unable to delete User")
		logging.Error("%s: %v", translation, err)
		http.Error(out, translation, http.StatusInternalServerError)
		return
	}
	formats.WriteJSONResponse(input, out, in)
}
